openlibrary/catalog/add_book/__init__.py

```diff
@@ -678,6 +684,70 @@ def load_data(rec, account_key=None):
     return reply
 
 
+def overwrite_promise_item_using_marc_record(
```

Comment:
Does this need to be so special cased?

It looks like we only ever update `number_of_pages` if the matched edition was an unedited (i.e v=1) promise item, and the new record  has a MARC record.

This won't support adding `number_of_pages` in the following cases

* user created edition with no `number_of_pages`
* user edited original promise item import (v > 1)

* any non-MARC import which has `number_of_pages` won't add it to an existing record even if it is needed.

Wouldn't it be better to update the standard edition modification code to do the best thing with the available import fields?

I don't see why the exact import source of the incoming record vs. existing record should make that much of a difference. 

Preferring existing values over new ones seems a sensible default, unless there is a good way to determine that a value is bad.

Reply:
I'll have a go at doing the best things with the available fields for the matched edition, and also preferring the existing values over the defaults, with perhaps an exception for a `publish_date` of `????`.

Reply:
@scottbarnes That sounds good, from having looked over this code just now, it looks like:

https://github.com/internetarchive/openlibrary/blob/3ef1957dbb4d30f66892a6c3a55c1f812b64a4ce/openlibrary/catalog/add_book/__init__.py#L841-L917

is the existing `modified` code, where we've found an existing match, and we are updating a selected number of fields.

That whole run of code looks like it could be split into two functions, one for `need_work_save` and one for `need_edition_save`, and all the correct checks and `modified` updates can be grouped sensibly.

I've worked on this code are before, and hadn't noticed that it was that badly in need of a refactor for clarity :)

Reply:
> Preferring existing values over new ones seems a sensible default, unless there is a good way to determine that a value is bad.

@hornc Anything from BWB or Amazon should be overwritten with prejudice by MARC data.

Reply:
This PR makes it so that BWB promise item data is overwritten by MARC data. It's a little difficult to detect this in the general case, but this handles where the _only_ on the record is from BWB promise items.